Studio Tagger

Flashy78 opened the pull request:
Resolves #3460 

Adds all of the options from #3460 with the tagger code copied over from the Performer tagger.
Since the Performer backend was refactored recently, I ended up refactoring the Studio backend to match. I found it was easier to copy how Performer was done rather than try to copy the tagger frontend and squish it into the existing backend.

Tested identify, import/export, scene tagger, studio tagger, studio tagger batch functions.. all with or without accompanying parent studio.

Flashy78 commented on go.mod:
```diff
@@ -11,6 +11,7 @@ require (
 	github.com/disintegration/imaging v1.6.0
 	github.com/fvbommel/sortorder v1.0.2
 	github.com/go-chi/chi v4.0.2+incompatible
+	github.com/gofrs/uuid v4.4.0+incompatible
```
This allows for passing in a search query that can be a name or a stash id, and parse the text to find out which one.

Flashy78 commented on internal/desktop/systray_nonlinux.go:
```diff
@@ -54,8 +57,9 @@ func systrayInitialize(shutdownHandler ShutdownHandler, faviconProvider FaviconP
 	c := config.GetInstance()
 	if !c.IsNewSystem() {
 		menuItems = c.GetMenuItems()
+		caser := cases.Title(language.English)
```
This was a linter failing.

WithoutPants commented on graphql/schema/types/scraper.graphql:
```diff
@@ -208,16 +223,18 @@ type StashBoxFingerprint {
-  "Refresh performers already tagged by StashBox if true. Only tag performers with no StashBox tagging if false"
+  "Refresh items already tagged by StashBox if true. Only tag items with no StashBox tagging if false"
   refresh: Boolean!
-  "If set, only tag these performer ids"
-  performer_ids: [ID!]
-  "If set, only tag these performer names"
-  performer_names: [String!]
```
We should not remove fields from inputs as this will break existing clients. We should mark these fields as deprecated and for now treat them as `ids` and `names` inputs.

Flashy78 replied:
Understood, will put them back in.

WithoutPants reviewed (changes requested):
Code review is still in progress.

Flashy78 commented on graphql/schema/types/scraper.graphql:
```diff
@@ -48,11 +48,21 @@ type ScrapedStudio {
   remote_site_id: String
 }
 
+type ScrapedParentStudio {
+    stored_id: ID
+    name: String!
+    url: String
+    image: String
+
+    remote_site_id: String
+}
```
This was my initial attempt before getting the parent value in ScrapedStudio to also be a ScrapedStudio.
Will remove.

Flashy78 commented on internal/api/resolver_model_studio.go:
```diff
@@ -2,30 +2,14 @@ package api
-	if obj.Name.Valid {
-		return obj.Name.String, nil
-	}
-	panic("null name") // TODO make name required
```
Name is required in the UI, but I don't think it's enforced anywhere. I will make a note to address that later.

Flashy78 commented on internal/api/resolver_mutation_studio.go:
```diff
@@ -28,190 +24,284 @@ func (r *mutationResolver) getStudio(ctx context.Context, id int) (ret *models.S
 	// Populate a new studio from the input
-	currentTime := time.Now()
 	newStudio := models.Studio{
-		Checksum:  checksum,
```
I moved checksum and the date fields into model_studio.go NewStudio and NewStudioPartial to avoid duplicating it everywhere, but looking now that doesn't stop anyone from just making a new object manually without using those methods and missing out on the fields. I will review this further.

Flashy78 replied:
Confirmed CreatedAt/UpdatedAt are enforced at the sqlite level.

Flashy78 commented on internal/api/resolver_mutation_studio.go:
```diff
@@ -28,190 +24,284 @@ func (r *mutationResolver) getStudio(ctx context.Context, id int) (ret *models.S
 	if input.IgnoreAutoTag != nil {
 		newStudio.IgnoreAutoTag = *input.IgnoreAutoTag
 	}
 
```
Image is now downloaded, but the saving to the db has been moved to pkg/sqlite/studio.go to avoid that duplicated block of code.
Same with saving the alias and stashid code blocks.

Flashy78 commented on internal/api/resolver_mutation_studio.go:
```diff
@@ -28,190 +24,284 @@ func (r *mutationResolver) getStudio(ctx context.Context, id int) (ret *models.S
+// This is slightly different to studioPartialFromStudioCreateInput in that Name is handled differently
+// and ImageIncluded is not hardcoded to true
+func studioPartialFromStudioUpdateInput(ctx context.Context, input StudioUpdateInput, id *string, translator changesetTranslator) (*models.StudioPartial, error) {
+	// Populate studio from the input
```
Probably a better way to avoid duplicating 98% of this code block.

Flashy78 commented on internal/identify/studio.go:
```diff
@@ -2,51 +2,128 @@ package identify
+
+// Duplicated in task_stash_box_tag.go
+func studioFromScrapedStudio(ctx context.Context, input *models.ScrapedStudio, endpoint string) (*models.Studio, error) {
+	// Populate a new studio from the input
```
I wasn't able to find a good way to make this method and the next reusable in both places, hence the duplication.

Flashy78 commented on internal/identify/studio_test.go:
```diff
@@ -87,18 +105,6 @@ func Test_createMissingStudio(t *testing.T) {
-			},
-			nil,
-			true,
-		},
```
With the removal of the UpdateStashIDs method, this test isn't really applicable.

Flashy78 commented on internal/manager/manager_tasks.go:
```diff
@@ -321,28 +323,38 @@ func (s *Manager) StashBoxBatchPerformerTag(ctx context.Context, input StashBoxB
+									excluded_fields: input.ExcludeFields,
+									task_type:       Performer,
+								})
+							}
```
Small refactor to load the StashIDs here instead of in the individual tasks. This way instead of adding a task for every single item and then deciding in the task if the user wanted to refresh existing or search for untagged, we are filtering here and only creating tasks for those items.

Flashy78 commented on internal/manager/manager_tasks.go:
```diff
@@ -418,3 +440,132 @@ func (s *Manager) StashBoxBatchPerformerTag(ctx context.Context, input StashBoxB
 }
+
+func (s *Manager) StashBoxBatchStudioTag(ctx context.Context, input StashBoxBatchTagInput) int {
+	j := job.MakeJobExec(func(ctx context.Context, progress *job.Progress) {
```
Almost a straight up duplication of the above method. I would like to refactor them in a future update.

Flashy78 commented on internal/manager/task_stash_box_tag.go:
```diff
@@ -51,7 +78,14 @@ func (t *StashBoxPerformerTagTask) stashBoxPerformerTag(ctx context.Context) {
+					return err
+				}
+			}
+			stashids := t.performer.StashIDs.List()
```
The StashIDs are now passed in to the task, but check to load them just in case.

Flashy78 commented on internal/manager/task_stash_box_tag.go:
```diff
@@ -135,6 +169,7 @@ func (t *StashBoxPerformerTagTask) stashBoxPerformerTag(ctx context.Context) {
+			//TODO: This seems incorrect, is there a reason it's opposite?
 			if excluded["name"] && performer.Name != nil {
 				partial.Name = models.NewOptionalString(*performer.Name)
 			}
```
This seems deliberate to have opposite logic for the excluded check, but I couldn't figure out what.
I did not replicate it for studio.

Flashy78 commented on internal/manager/task_stash_box_tag.go:
```diff
@@ -258,6 +293,260 @@ func (t *StashBoxPerformerTagTask) stashBoxPerformerTag(ctx context.Context) {
+					r.hookExecutor.ExecutePostHooks(ctx, *updatedStudio.ParentID, plugin.StudioUpdatePost, input, parentTranslator.getFields())
+				}
+				r.hookExecutor.ExecutePostHooks(ctx, updatedStudio.ID, plugin.StudioUpdatePost, input, translator.getFields())
+			*/
```
Wondering if the hooks should be run during batch operations.
If so, they will need to refactored to be available to this package somehow.

Flashy78 commented on pkg/models/model_scraped_item.go:
```diff
@@ -6,15 +6,28 @@ import (
 
 func (ScrapedStudio) IsScrapedContent() {}
 
+type ScrapedParentStudio struct {
```
Will remove.

Flashy78 commented on pkg/models/model_studio.go:
```diff
@@ -56,3 +101,12 @@ func (s *Studios) Append(o interface{}) {
+
+	ParentCreate *Studio
+	ParentUpdate *StudioPartial
+}
```
This is definitely a different architecture than currently exists, but for calls to Create I needed a StudioCreate, and maybe a ParentCreate or ParentUpdate. For calls to UpdatePartial, I needed a StudioUpdate and maybe a ParentCreate or ParentUpdate.
I tried a few different things like passing them in as params, or stuffing the parent objects into the studio objects, and passing flags for what to create/update... but this seemed to work the best.

WithoutPants replied:
See my comment in the graphql changes. The operation needs to be split up. Even if we disregard my comments there, I don't think this is a good solution to what you're trying to do. The `sqlite` code shouldn't be responsible for updating related objects; that's the job of the calling code.

Flashy78 commented on pkg/models/studio.go:
```diff
@@ -56,19 +57,17 @@ type StudioReader interface {
-	UpdateFull(ctx context.Context, updatedStudio Studio) (*Studio, error)
+	Create(ctx context.Context, input StudioDBInput) (*int, error)
+	UpdatePartial(ctx context.Context, input StudioDBInput) (*Studio, error)
+	Update(ctx context.Context, updatedStudio *Studio) error
```
Updated methods to match Performer, except Create returns the created id instead of accepting the studio as a pointer and updating in place. I've been hurt too many times by that pattern of not realizing my object is getting updated inside a method without me knowing, so I try to avoid it.

Flashy78 replied:
Reverted my changed for Create.
Should I revert UpdatePartial as well? What's the use case for passing in the id separate instead of as part of the partial model being updated?

Flashy78 commented on pkg/models/timestamp_test.go:
```diff
@@ -37,7 +37,7 @@ func TestTimestamp(t *testing.T) {
 		{"date", "2021-04-05", "2021-04-05T00:00:00Z"},
 		{"datetime", "2021-04-05 14:45:36", "2021-04-05T14:45:36Z"},
-		{"datetime-tz", "2021-04-05 14:45:36 PDT", "2021-04-05T14:45:36Z"},
+		{"datetime-tz", "2021-04-05 14:45:36 PDT", "2021-04-05T14:45:36-07:00"},
```
This test was strangely failing for me, but I assume must have worked for previous PRs. At least my change also seems to pass the Github build.

Flashy78 replied:
Ok, it fails on Github, will revert.

Flashy78 commented on pkg/sqlite/repository.go:
```diff
@@ -516,21 +516,6 @@ func (r *stashIDRepository) get(ctx context.Context, id int) ([]models.StashID,
-	}
-	return nil
-}
-
```
I guess Studio must have been the only one using this method as the linter wanted me to remove it.

Flashy78 commented on pkg/studio/export_test.go:
```diff
@@ -181,26 +172,13 @@ func TestToJSON(t *testing.T) {
-	mockStudioReader.On("GetStashIDs", ctx, studioID).Return(stashIDs, nil).Once()
-	mockStudioReader.On("GetStashIDs", ctx, noImageID).Return(nil, nil).Once()
-	mockStudioReader.On("GetStashIDs", ctx, missingParentStudioID).Return(stashIDs, nil).Once()
-
```
No need to test these as they were removed.

Flashy78 commented on pkg/studio/import.go:
```diff
@@ -113,16 +98,6 @@ func (i *Importer) PostImport(ctx context.Context, id int) error {
-	if err := i.ReaderWriter.UpdateAliases(ctx, id, i.Input.Aliases); err != nil {
-		return fmt.Errorf("error setting tag aliases: %v", err)
-	}
-
```
These were moved to pkg/sqlite/studio.go, but creating the image was left here.

Flashy78 commented on ui/v2.5/src/components/Dialogs/IdentifyDialog/constants.ts:
```diff
@@ -31,6 +31,8 @@ export const multiValueSceneFields: SceneField[] = [
   if (field === "code") {
     return "scene_code";
+  } else if (field === "studio") {
+    return "studio_and_parent";
```
Defaulted to also creating parents when the user has chosen to create studios.

Flashy78 commented on ui/v2.5/src/components/Tagger/performers/PerformerTagger.tsx:
```diff
@@ -418,7 +418,7 @@ const PerformerTaggerList: React.FC<IPerformerTaggerListProps> = ({
             name="performer-query"
             label={<FormattedMessage id="performer_tagger.current_page" />}
-            defaultChecked
+            defaultChecked={!queryAll}
```
Fix so the modal remembers what has been checked.

Flashy78 commented on ui/v2.5/src/components/Tagger/queries.ts:
```diff
@@ -204,6 +208,54 @@ export const useUpdateStudioStashID = () => {
+
+        if (updatedStudio.data?.studioUpdate?.parent_studio) {
+          const ac = getClient();
+          ac.refetchQueries({ include: studioMutationImpactedQueries });
```
The goal of this was that if a parent was created/updated, since the data for that studio was not being passed back to update the cache directly, the whole thing should be refreshed.
Perhaps there is a better way of doing this.

Flashy78 commented on ui/v2.5/src/components/Tagger/scenes/StudioModal.tsx:
```diff
@@ -1,80 +1,236 @@
   const renderField = (
     id: string,
     text: string | null | undefined,
-    truncate: boolean = true
```
I think I should put this back in.

Flashy78 commented on ui/v2.5/src/components/Tagger/studios/StudioTagger.tsx:
```diff
@@ -0,0 +1,750 @@
+
+      // Once the studio batch is complete, refresh all local studio data
+      const ac = getClient();
+      ac.refetchQueries({ include: studioMutationImpactedQueries });
```
Again the goal is to refresh the local studio cache, not sure if this is the best way to do it.

WithoutPants commented:
The sqlite refactoring is now merged, so this can be updated when you're ready.

DingDongSoLong4 commented on pkg/sqlite/studio.go:
```diff
@@ -183,13 +291,16 @@ func (qb *StudioStore) Destroy(ctx context.Context, id int) error {
 
-// returns nil, nil if not found
+// returns nil, sql.ErrNoRows if not found
 func (qb *StudioStore) Find(ctx context.Context, id int) (*models.Studio, error) {
```
The reason why I intentionally split the Find and find methods in my refactoring is to prevent sql.ErrNoRows errors from escaping out from the sqlite package. External code (such as resolvers) use the Find method, meaning that a graphql query for a missing studio returns null instead of an error. Internal code can use either method depending on whether the id is valid or not should be considered an error.

Flashy78 replied:
Understood, sorry I didn't catch that reasoning from the code. I'll add it back in.

DingDongSoLong4 reviewed (commented):
Ultimately it's not up to me to decide how these things are structured, but the way I see the `sqlite` package is that it should be a simple interface around the database, generating SQL to insert/update/delete records as required by the rest of the system. In other words, logic for things like ensuring parent studios, setting updated at timestamps, etc is a better fit for `pkg/studio` rather than `pkg/sqlite`.

The motivation for my sqlite refactoring, and for some of the other refactorings I have lined up, is that I want to eventually try and add support for other database engines (specifically, postgres). Having all entity create/update/delete actions first go through validation in their respective `pkg/{scene,image,studio,etc}` packages (ie the `Service` structs for scenes, images and galleries) rather than using the methods in `pkg/sqlite` directly would make it much easier for different database backends to exist and would also ensure invalid data never actually gets stored.

However that is all just my opinion, so until a clear structure is worked out nothing really has to change here.

Flashy78 commented:
> Ultimately it's not up to me to decide how these things are structured, but the way I see the `sqlite` package is that it should be a simple interface around the database, generating SQL to insert/update/delete records as required by the rest of the system. In other words, logic for things like ensuring parent studios, setting updated at timestamps, etc is a better fit for `pkg/studio` rather than `pkg/sqlite`.
> 
> The motivation for my sqlite refactoring, and for some of the other refactorings I have lined up, is that I want to eventually try and add support for other database engines (specifically, postgres). Having all entity create/update/delete actions first go through validation in their respective `pkg/{scene,image,studio,etc}` packages (ie the `Service` structs for scenes, images and galleries) rather than using the methods in `pkg/sqlite` directly would make it much easier for different database backends to exist and would also ensure invalid data never actually gets stored.
> 
> However that is all just my opinion, so until a clear structure is worked out nothing really has to change here.

That does make sense, but as you say I think there needs to be a slight rearchitecture of how the sqlite package is used in order to support that design.

My reasoning is because there is no clear structure, timestamps were set in some places and not in others. If someone used the models.NewPartialObject() method they got timestamps, if they just instantiated a new object on their own then they didn't. So enforcing those at the lowest level ensured they were always set correctly and the developer did not have to always remember to do it. It's even easier if those fields are set directly in the database.

There are so many places were logic is just duplicated everywhere, like saving the image for an object. Where instead if the logic is at the lowest point in the chain, developers don't have to remember to copy and paste that block of code every time.

I think if there was a way to stop developers from using the sqlite package directly, then all that logic could go in the layer above it, but it doesn't seem like things are architected that way currently.

DingDongSoLong4 commented:
Agreed. Like I said, changing anything now isn't that necessary, as long as everything works. The reorganising can come later, I'll get to it eventually.

WithoutPants commented on internal/api/resolver_query_find_studio.go:
```diff
@@ -8,6 +8,10 @@ import (
 )
 
 func (r *queryResolver) FindStudio(ctx context.Context, id string) (ret *models.Studio, err error) {
+	if id == "" {
+		return nil, nil
+	}
```
This is not consistent with the behaviour of the other resolvers. An empty id should be an error condition.

WithoutPants commented on graphql/schema/types/studio.graphql:
```diff
@@ -29,6 +29,7 @@ input StudioCreateInput {
   name: String!
   url: String
   parent_id: ID
+  parent: StudioCreateInput
```
Looking at the way this is processed in the resolver logic, I don't think it's a good design to include this. The behaviour isn't clear from the outside and the behaviour is based on whether an existing studio exists or not. The client should be responsible for creating/updating the parent studio. The only benefit I see in combining these operations is that you end up with a single transaction to roll back if needed, but I don't think it's worth it.

Flashy78 replied:
I've moved all parent studio logic out of the `sqlite` package.
I think I was trying to treat it logically as a single operation (not necessarily a single transaction). I wanted to avoid duplicating all of the parent matching/creating/updating logic in all of the multiple places where studios are created (identify/import/batch update/tagger view).
All of that code duplication can lead to bugs in the future when there's a new way to create a studio but all of the parent logic is not copied exactly. Like when disambiguation was added, it was added to all of the various places you can create a performer, but missed in identify.

I think this comment means you're suggesting to also write this parent logic in the frontend, so that the frontend can make an api call to create/update a parent studio, then attach it to the main studio and make another api call to create/update studio?

WithoutPants reviewed (changes requested):
This is not a complete review as there's a significant design issue that I think needs to be addressed before I continue - regarding the `parent` field of the studio create/update mutation. Happy to discuss this further as needed.

Flashy78 commented on internal/manager/task_stash_box_tag.go:
```diff
@@ -183,7 +210,261 @@ func (t *StashBoxPerformerTagTask) stashBoxPerformerTag(ctx context.Context) {
+				StashID:  *input.RemoteSiteID,
+			},
+		},
+		Mode: models.RelationshipUpdateModeSet,
```
This looks a bit ugly now because I first have to load the existing studio loaded in order to load the existing ids.

It feels like there should be a way to just add new/updated ids and then the sqlite UpdatePartial method knows it's doing a partial update and not to replace all ids.

Flashy78 commented on internal/api/resolver_model_studio.go:
```diff
@@ -99,7 +99,14 @@ func (r *studioResolver) ParentStudio(ctx context.Context, obj *models.Studio) (
 	}
 
-	return loaders.From(ctx).StudioByID.Load(*obj.ParentID)
+	if err := r.withReadTxn(ctx, func(ctx context.Context) error {
+		ret, err = r.repository.Studio.Find(ctx, *obj.ParentID)
+		return err
+	}); err != nil {
+		return nil, err
+	}
+
+	return ret, nil
```
I don't remember, reverted.

Flashy78 commented on internal/manager/task_stash_box_tag.go:
```diff
@@ -183,7 +210,261 @@ func (t *StashBoxPerformerTagTask) stashBoxPerformerTag(ctx context.Context) {
+			}
+		}
+	} else {
+		partial.ParentID = models.NewOptionalIntPtr(nil)
```
This code is now in `model_scraped_item.go`.
I think it's kind of just like the URL or image field. If you've said you want to batch update all of your studios with Stash-box data and haven't excluded that field, then it gets set to whatever is scraped.

WithoutPants commented:
I've made a bunch of changes to the backend code.

I've removed the `parent` field from the create and update inputs. The front end code needs to be updated to separately create/update parent studios. Currently, the tagger hangs when creating a studio with parent info, because the server is returning an error and the error isn't being handled correctly. I'll investigate this myself tomorrow if you don't get to it in the meantime.

I've taken out the image data from the studio object and partial structs. It was a vestigial field on the studio struct, only used during creation. It's also inconsistent with the other objects/partials. It does result in more code on the calling side to set the image, but this can be mitigated in future by combining the create/update object and image code into a service-level function. I could have kept the image in the partial, but this would be inconsistent with the create behaviour.

I reverted some undocumented design decisions such as moving the timestamp setting code into the sqlite package. In principle, the sqlite package should be concerned only with querying and modifying data, and not with the logic. We want fine control of these timestamps outside the data layer, and as such should not have been moved here. The alias uniqueness checking should probably not be in sqlite either, but for now it's fine to stay there. In future it should be moved into the business layer. In future we should unify the create/update functions into a single function in the studio package, which would include this validation.

I also reverted a change that prevented aliases from being used in studio autotag. This looked like it was copied from the performer code, but the reasons for disabling aliases in the performer code do not apply to studios.

In future, I'd prefer if you could separate large refactors from new features. It makes reviewing far more difficult. In future, please also highlight low-level design decisions like the items mentioned above. These have far-reaching impacts on the design and behaviour, and detecting them with static code reviews are difficult.

Flashy78 commented:
> I've made a bunch of changes to the backend code.
> 
> In future, I'd prefer if you could separate large refactors from new features. It makes reviewing far more difficult. In future, please also highlight low-level design decisions like the items mentioned above. These have far-reaching impacts on the design and behaviour, and detecting them with static code reviews are difficult.

Thanks for taking the time to get this where it needs to be.

I will avoid new features and stick to bug fixes to avoid creating more work for you.

WithoutPants commented:
> I will avoid new features and stick to bug fixes to avoid creating more work for you.

This wasn't intended to discourage you from making new features! The overall change here is all good, there was just some issues around the refactoring that was done at the same time.

WithoutPants commented:
I've done some refactoring of the UI code to handle creating/updating parent studios. I'd appreciate if you could give it a test run to ensure I haven't broken anything. It's looking about complete on my end.

Flashy78 commented:
I was too overconfident in my ability to pick up and understand go. I will lurk longer to learn more about how the codebase is structured/designed (ie sqlite should only read/write with zero changes to the objects). I would rather write features that you can bug check/suggest changes, rather than having to spend days rewriting. In that case it would be more efficient just for you to write the feature whenever it appears on your list.

If I tag a studio, where the parent studio exists but has no stashId, it tries to create the parent instead of updating.
```
{
    "errors": [
        {
            "message": "inserting into studios: executing `INSERT INTO `studios`....: UNIQUE constraint failed: studios.name",
            "path": [
                "studioCreate"
            ]
        }
    ],
    "data": {
        "studioCreate": null
    }
}
```

Some cache seems to be getting messed up, which may be the cause above. I tried to reproduce and it wanted to create the parent. Refreshed the page and it correctly matched the parent and assigned the stashid.

Deleted stashId from Studio, deleted parent studio, navigate back to tagger, studio modal thinks the parent still exists and then 
```
{
    "errors": [
        {
            "message": "studio with id 30 not found",
            "path": [
                "studioUpdate"
            ]
        }
    ],
    "data": {
        "studioUpdate": null
    }
}
```
